### PR TITLE
Add ScreenClassContext + ScreenClassProvider to attach screenClass updates to one resize handler

### PR DESCRIPTION
**Pull request by LMulvey**
Addresses #81 

This is my shot at implementing context to use one source of truth for the `getScreenClass` function calls, as well as reduce the amount of re-renders during resizing from ~2200 to 8 (resizing from xl -> xs -> xl).

**React Profiler Flamechart with ScreenClassContext**
image

**React Profiler Flamechart _without_ ScreenClassContext**
image

There is, of course, downsides to this method and while I feel that it should be absolutely explored by the maintainers, it may not be ready for merge _yet_. The largest downside is that it introduces a **breaking change**: all internal components rely on that context to provide them the current screenClass–if your React application is _not_ wrapped in a ScreenClassProvider, you'll receive `screenClass is null` errors. 

Implementation of this can be simple in basic apps, but, with larger applications, it may not be so simple (depending on how you're using the components). I had fiddled with implementing them directly into the `Container` components but wasn't ready to make that assumption that all users wrap their `Row` and `Col` components in a `Container`.

Basic implementation looks something like this:

```
import { Container, Row, Col, ScreenClassProvider } from 'react-grid-system';

export default function App() {
  return (
    <ScreenClassProvider>
      <div className="App">
           {children}
      </div>
    </ScreenClassProvider>
```

Of course, if your app is more complex, you may need to change your implementation. Especially if you're doing something like developing an internal component library with Storybook then you need to find a way to gracefully add that provider to all components that require it. 

TL;DR, huge performance benefits from propagating the screenClass in this way but it has drawbacks that should be explored fully.

### CODE REVIEW

**Comment by CharlyJazz** [comment]
Wow that is amazing @tikider 
Ready to merged?

**Comment by gerbenmeyer** [comment]
Cool stuff! I would be in favor though to solve this in a non breaking way, so e.g. make the ScreenClassProvider part of Container, and let Col fall back to the old approach if the screenClass is null.

**Comment by LMulvey** [comment]
I wonder if it's possible to have it both ways–allow a user to have a top-level provider to handle the context, or have it at a container level, but also have it in Col/ScreenClassRender/Hidden/Visible for fallbacks? Allows flexibility based on everyone's own implementation.

I had opted to not use it directly in the container as in the project we're currently using this in, we are using a large amount of Containers, which probably _isn't_ ideal. 

I can certainly put a bit more work into this to make it not break any currently implementation and make upgrading painless for current users and have global providers be opt-in.

**Comment by LMulvey** [comment]
To update, the solution I've come to that would be the most fair and least-API-breaking:

- Fallback provider in `Container`
- Fallback providers inside of the helper components (`Visible`, `Hidden`, and `ScreenClassRender`) that will rarely be wrapped in `Container` components.
- Fallback inside of Col/Row if there is no provider to use it's own ScreenClass source.

This will allow all current implementations of the package to continue on gracefully but allow larger projects the option to have a top-level Provider that would subscribe all lower components to that _one_ `ScreenClass` as the source of truth, as well as one resize handler.

Those with the current implementation will still see a performance boost if they're using a small amount of `Container` components to structure their layouts. I'd say it's a win all around! I'll get to work on a PR but let me know if you have any thoughts in the meantime - @gerbenmeyer @CharlyJazz

**Comment by CharlyJazz** [comment]
News?

**Comment by LMulvey** [comment]
Still on my radar! To spare you the details of my personal life, I've offloaded one of my part-time commitments after this week and will have more time to contribute to things I enjoy (like this), so I should have something to show shortly.

**Comment by CharlyJazz** [comment]
Can you make a list of tasks to be done? Maybe we can help you! We should talk about migrating to React 16.6

**Comment by LMulvey** [comment]
Hey! Sorry for the wait between pushes here. So, updates:
1) `ScreenClassContext` is initialized with a `defaultValue` of `'NO_PROVIDER'`
2) At a component-level, I do some checking to see if a provider exists. If it doesn't, the render prop's only argument will be `'NO_PROVIDER'`.
3) If there is no provider, I provide one.
4) I've abstracted this logic to the `utilities/ScreenClassResolver` component so the `render` for Row and Col doesn't get out of hand.

I did some testing and the performance benefits are still present IF you have your own provider, otherwise, it works just the same as it always did–so no API breaking changes. 

I've still gotta add this functionality to the Hidden and ScreenClassRender components but wanted to confirm that this is headed in the right direction. @gerbenmeyer @CharlyJazz

**Comment by LMulvey** [comment]
Did a little more cleanup here as well. Once you're good with the changes, I can put together a bit of documentation for it if needed.

**Comment by gerbenmeyer** [comment]
Nice stuff, definitely good with this :smiley: 

Once you've updated the documentation as well, I'll merge it in and release a new version.

**Comment by LMulvey** [comment]
Cool! I think I covered everything off. Also caught a typo in 4a7c754 that would have caused a memory leak 😄

**Comment by LMulvey** [comment]
@gerbenmeyer - Heyo! I'm wondering if there's anything else you need me to complete here or if it's GTM at this stage 😄

**Comment by gerbenmeyer** [comment]
Ah yes, let's merge it :smiley:

**Comment by gerbenmeyer** [comment]
And it's published to npm as well! :tada:

**Comment by LMulvey** [comment]
Glad to hear! I'll keep an eye on the issues for any weird bugs but we implemented it this morning in our app and it's working great 👍
